### ISSUE TITLE
Streamline views into observed styled-components

### ISSUE THREAD

**Issue by andraz** [mobxjs/mobx-react#535]
Maybe we should continue debating https://github.com/mobxjs/mobx-state-tree/issues/988 here?

https://github.com/andraz/mobx-styled-observers

demo: https://codesandbox.io/s/github/andraz/mobx-styled-observers/tree/master/

Where does the `observers` HOC belong into the MobX ecosystem and how could we improve some more on this pattern? 

Point of this debate is:
- to join all connected concerns 
 `action -> update data -> recalculate -> view -> component -> <Jsx onClick={action} />`
 on the same spot while moving all concerns which do not belong in _source of truth_ (css styling, etc ...) to external imports
- to reduce the barrier to entry for newbies (programming and designers) as much as possible
- eliminate big area prone to advanced-level subtle bugs by completely automating it
  - passing of stores via props
  - injecting stores with decorators
  - manual wrapping JSX with observers
  - reducing unneeded repaints (observed component avoids parent repaint)

use case:
 ```
// Button = imported StyledComponent (component with built in `css`)
// todo.textButtonText = calculated view used to generate button's label

.views(todo =>
    observers({ // wrap all "render-views" with a custom HOC
      TextButton: () => (  // component gets render(), inject('store') etc, all built in
        <Button onClick={todo.toggle}>{todo.textButtonText}</Button>
      ),
      CountButton: () => ( // is used without props or decorators: <CountButton />
        <Button onClick={todo.toggle}>{todo.countButtonText}</Button>
      )
    })
  )
```

**Comment by andraz** [comment]
I have cleaned up some of the code to simplify the use case.

Now the view is used only as a getter to provide a cached closure for store. The HOC now lands outside of the MST files.

```
import button from './button.component'
const { CountButton, TextButton } = button

...

    get TextButton() {
      // wrap store into the returned styled component via closure
      return () => <TextButton todo={todo} />
    },
    get CountButton() {
      // getter can use same name as the component
      return () => <CountButton todo={todo} /> // component name
    }
```

/src/button.component.js
```
import React from 'react'
import Button from './button.style'
import { observers } from './util'

const TextButton = ({ todo }) => (
  <Button onClick={todo.toggle}>{todo.textButtonText}</Button>
)

const CountButton = ({ todo }) => (
  <Button onClick={todo.toggle}>{todo.countButtonText}</Button>
)

export default observers({ TextButton, CountButton })
```

Can the `observers` HOC get added into mobx-react ?
It is a tiny wrapper for `observer`, but would be useful if support for objects came already built in.

**Comment by mweststrate** [comment]
If I am correct the pattern is just `const observers = comps => lodash.map(comps, observer)`? To me that sounds like a syntactic sugar that is so trivial that there is no real need to make that part of the library. I prefer to keep things that can easily be expressed in user land outside the core libraries, as this kind of features, although useful, tend to blow up the api surface and thereby the cognitive load of new folks learning a library. 

Beyond that I think most people stick to one component per file, in which case `observers` couldn't be used anyway. (If I understand correctly what you are proposing)

**Comment by andraz** [comment]
I developed this idea out of an experiment you made on codesandbox.io on wrapping multiple actions so they return observable components.

The gist of this new pattern is to hide all "injecting and props passing" behind closures inside actions. This causes them to simply "work" out of the box, with high performance optimal repaints, virtually "without wiring", when simply "thrown anywhere" provided by store getters. Now this is a noobproof concept... ;) 

Apparently we now also both came to the same conclusion about the non-need for HOC at about the same time.

I recently deployed new version of [mobx-styled-observers](https://codesandbox.io/s/github/andraz/mobx-styled-observers) software design pattern without it.

Latest version simplifies code so it does not need the HOC because refactoring to using a well known folder layout + splitting components to files, greatly reduces any need to wrap multiple components. Object literals would make code less DRY, so I threw that out and with it any need for HOC.

I still need to do some more research on nesting and how to optimize the wiring in MST `map()` or `array()` to make looping through all children (to get render components from them) as neat as possible.

Tnx for support.
